File: 2. 两数相加.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    const dummyHead = new ListNode();
    let ptr = dummyHead;
    let increment = 0;
    while (l1 || l2) {
        const val1 = l1 ? l1.val : 0;
        const val2 = l2 ? l2.val : 0;
        let value = val1 + val2 + increment;
        increment = Math.floor(value / 10);
        value = value % 10;

        ptr.next = new ListNode(value);
        ptr = ptr.next;
        if (l1) l1 = l1.next;
        if (l2) l2 = l2.next;
    }
    if (increment) {
        ptr.next = new ListNode(increment);
    }
    return dummyHead.next;
};

File: 2. 两数相加.cpp
/**
 * Definition for singly-linked list->
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* dummyHead = new ListNode();
        ListNode* ptr = dummyHead;
        int increment = 0;
        while (l1 && l2) {
            int value = l1->val + l2->val + increment;
            increment = value / 10;
            value %= 10;

            ptr->next = new ListNode(value);
            ptr = ptr->next;
            l1 = l1->next;
            l2 = l2->next;
        }
        while (l1) {
            int value = l1->val + increment;
            increment = value / 10;
            value %= 10;
            ptr->next = new ListNode(value);
            ptr = ptr->next;
            l1 = l1->next;
        }
        while (l2) {
            int value = l2->val + increment;
            increment = value / 10;
            value %= 10;
            ptr->next = new ListNode(value);
            ptr = ptr->next;
            l2 = l2->next;
        }
        if (increment) {
            ptr->next = new ListNode(increment);
        }
        return dummyHead->next;
    }
};